ecommerce/extensions/payment/processors/payu.py

```diff
@@ -0,0 +1,292 @@
+            response (dict): Dictionary of parameters received from the payment processor.
+
+        Keyword Arguments:
+            basket (Basket): Basket being purchased via the payment processor.
+
+        Raises:
+            TransactionDeclined: Indicates the payment was declined by the processor.
+            GatewayError: Indicates a general error on the part of the processor.
+            InvalidCyberSourceDecision: Indicates an unknown decision value
+        """
+
+        # Validate the signature
+        if not self.is_signature_valid(response):
+            raise InvalidSignatureError
+
+        # Raise an exception for payments that were not accepted. Consuming code should be responsible for handling
+        # and logging the exception.
+        decision = response['state_pol']
+        if decision != '4':
+            exception = {
+                '6': TransactionDeclined,
+                '104': GatewayError
+            }.get(decision, InvalidPayUDecision)
+
+            raise exception
+
+        # Create Source to track all transactions related to this processor and order
+        source_type, __ = SourceType.objects.get_or_create(name=self.NAME)
+        currency = response['currency']
+        total = Decimal(response['value'])
+        transaction_id = response['transaction_id']
+        req_card_number = response['cc_number'] or ''
+        # card_type = CYBERSOURCE_CARD_TYPE_MAP.get(response['req_card_type'])
+
+        source = Source(source_type=source_type,
+                        currency=currency,
+                        amount_allocated=total,
+                        amount_debited=total,
+                        reference=transaction_id,
+                        label=req_card_number)
+
+        # Create PaymentEvent to track
+        event_type, __ = PaymentEventType.objects.get_or_create(name=PaymentEventTypeName.PAID)
+        event = PaymentEvent(event_type=event_type, amount=total, reference=transaction_id, processor_name=self.NAME)
+
+        return source, event
+
+    def _generate_signature(self, parameters, signature_type=1):
+        """
+        Sign the contents of the provided transaction parameters dictionary.
+
+        This allows PayU to verify that the transaction parameters have not been tampered with
+        during transit.
+
+        We also use this signature to verify that the signature we get back from PayU is valid for
+        the parameters that they are giving to us.
+
+        Arguments:
+            parameters (dict): A dictionary of transaction parameters.
+
+        Returns:
+            unicode: the signature for the given parameters
+        """
+
+        # sigantures to validate payment form
+        if signature_type == 1:
+            uncoded = "{ApiKey}~{merchantId}~{referenceCode}~{amount}~{currency}".format(
+                ApiKey=self.ApiKey,
+                merchantId=self.merchantId,
+                referenceCode=parameters['referenceCode'],
```

Comment:
It is not possible, since we must set the value of `referenceCode` parameter before calling the payment processor form